types/Impl/Product.php

```diff
@@ -0,0 +1,40 @@
+<?php
+declare(strict_types=1);
+
+namespace Garp\Functional\Types\Impl;
```

Comment:
Do these classes actually need a separate namespace? And if they do, maybe create a new namespace: `Garp\Functional\Impl` (discussion about the name stil continues) besides `Garp\Functional\Types`.

Reply:
Well, I would just like to group them _somewhere_... they are distinctly different from everything else in this library. 
But it's a good idea to pull them out of `Types`. 👍